Fix #4422, use grad instead of grads[0] for cudnn_batch_norm derivative.

ezyang opened the pull request:
The fix proper is very simple: use 'grad' instead of 'grads[0]' in
cudnn_batch_norm's backward definition.

However, the problem was annoyingly subtle and I nearly went off
and implemented a big pile of infrastructure to "tell" the codegen
how to distinguish between differentiable and non-differentiable
outputs before realizing that there must be a way we do this
legitimately, because it works for THNN.  I documented this in
derivatives.yaml, and also added tests for the problem in
load_derivatives.py to catch the various ways you could "get it
wrong".  Hope this helps someone else.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

CC @elbaro

ezyang committed:
Fix #4422, use grad for cudnn_batch_norm derivative / don't use toTensor()

This commit fixes double-backwards on batch norm.  There were two
bugs:

- Returned buffers from batchnorm backwards were being marked as differentiable
  when they shouldn't be.  The fix for this is "easy": use 'grad' instead of
  'grads[0]' in cudnn_batch_norm's backward definition.  (More on this below.)

- I was using toTensor on a Scalar, which gives me a Tensor of the wrong
  type when I'm in CUDA world.  Using the Scalar add() overload directly
  solves the problem.

The differentiability of returned buffers was annoyingly subtle and I nearly
went off and implemented a big pile of infrastructure to "tell" the codegen how
to distinguish between differentiable and non-differentiable outputs before
realizing that there must be a way we do this legitimately, because it works for
THNN.  I documented this in derivatives.yaml, and also added tests for the
problem in load_derivatives.py to catch the various ways you could "get it
wrong".  Hope this helps someone else.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

colesbury commented on test/test_nn.py:
```diff
@@ -3233,6 +3233,13 @@ def func(root):
         gradgradcheck(func, [v])
 
+    def test_batchnorm_double_backward(self):
+        bn = nn.BatchNorm1d(5).cuda()
```
The CUDA calls should be guarded by a check or a `@unittest.skipIf`.

Are we doing gradgradchecks of BatchNorm? For both THNN and cuDNN?

ezyang replied:
Good catch, I forgot to reenable the tests I disabled >.>

apaszke commented on tools/autograd/load_derivatives.py:
```diff
@@ -129,6 +129,54 @@ def set_up_derivatives(defn, declaration):
             derivatives.append(create_derivative(declaration, formula, output_indices, args))
 
+        # Test to see if the use of 'grads' makes sense.
+        used_grad = 0
```
This is a large chunk of the outer function, can we move it out somewhere else with an appropriate name, so we don't make this one much less readable?

ezyang committed:
Print full type of Variable tensor

Previously, it printed [Variable]; now it prints [Variable CPUDoubleTensor].
I'm not altogether sure why toString on Variable returns the uninformative
thing, but that might be worth fixing too.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot retest this please
